File: implementations/java/common/src/main/java/com/theshuai/common/peermesh/PeerMdnsBrowser.java
package com.theshuai.common.peermesh;

import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Minimal DNS-SD browse for local HTTP/SSH/UDP services. Only loopback and local-interface
 * targets survive {@link PeerServiceDiscovery#requireTargetHost(String)}.
 */
public final class PeerMdnsBrowser {
    private static final String[] QUERIES = {
            "_http._tcp.local",
            "_https._tcp.local",
            "_ssh._tcp.local",
            "_udp.local"
    };

    private PeerMdnsBrowser() {
    }

    public static List<PeerMdnsCandidate> browse(Duration timeout) {
        List<byte[]> packets = new ArrayList<>();
        try (DatagramSocket socket = new DatagramSocket()) {
            socket.setSoTimeout((int) Math.max(50, timeout.toMillis() / Math.max(1, QUERIES.length)));
            socket.setBroadcast(true);
            InetAddress mdns = InetAddress.getByName("224.0.0.251");
            for (String name : QUERIES) {
                byte[] query = encodePtrQuery(name);
                socket.send(new DatagramPacket(query, query.length, mdns, 5353));
            }
            long deadline = System.currentTimeMillis() + Math.max(200, timeout.toMillis());
            byte[] buffer = new byte[1500];
            while (System.currentTimeMillis() < deadline) {
                DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
                try {
                    socket.receive(packet);
                    packets.add(java.util.Arrays.copyOf(packet.getData(), packet.getLength()));
                } catch (Exception ignored) {
                    break;
                }
            }
        } catch (Exception ignored) {
            return List.of();
        }
        return parseResponses(packets);
    }

    static List<PeerMdnsCandidate> parseResponses(List<byte[]> packets) {
        Map<String, PeerMdnsCandidate> found = new LinkedHashMap<>();
        Map<String, String> ptr = new LinkedHashMap<>();
        Map<String, Srv> srv = new LinkedHashMap<>();
        Map<String, String> addr = new LinkedHashMap<>();
        for (byte[] packet : packets) {
            parsePacket(packet, ptr, srv, addr);
        }
        for (Map.Entry<String, String> entry : ptr.entrySet()) {
            Srv record = srv.get(entry.getValue().toLowerCase(Locale.ROOT));
            if (record == null) {
                continue;
            }
            String host = addr.getOrDefault(record.target.toLowerCase(Locale.ROOT), record.target);
            try {
                host = PeerServiceDiscovery.requireTargetHost(host);
            } catch (RuntimeException ignored) {
                continue;
            }
            String application = applicationFor(entry.getKey());
            String transport = PeerServiceDiscovery.APPLICATION_UDP.equals(application)
                    ? PeerServiceDiscovery.TRANSPORT_UDP : PeerServiceDiscovery.TRANSPORT_TCP;
            PeerMdnsCandidate candidate = new PeerMdnsCandidate();
            candidate.setName(instanceName(entry.getValue(), application));
            candidate.setTransport(transport);
            candidate.setApplication(application);
            candidate.setTargetHost(host);
            candidate.setTargetPort(record.port);
            found.put(candidate.getName() + ":" + host + ":" + record.port, candidate);
        }
        return List.copyOf(found.values());
    }

    private static void parsePacket(byte[] packet,
                                    Map<String, String> ptr,
                                    Map<String, Srv> srv,
                                    Map<String, String> addr) {
        if (packet == null || packet.length < 12) {
            return;
        }
        ByteBuffer buffer = ByteBuffer.wrap(packet);
        buffer.position(4);
        int questions = Short.toUnsignedInt(buffer.getShort());
        int answers = Short.toUnsignedInt(buffer.getShort());
        int authority = Short.toUnsignedInt(buffer.getShort());
        int additional = Short.toUnsignedInt(buffer.getShort());
        for (int i = 0; i < questions && buffer.hasRemaining(); i++) {
            skipName(buffer);
            if (buffer.remaining() < 4) {
                return;
            }
            buffer.getShort();
            buffer.getShort();
        }
        int records = answers + authority + additional;
        for (int i = 0; i < records && buffer.remaining() >= 10; i++) {
            int nameAt = buffer.position();
            String name = readName(buffer, packet);
            if (buffer.remaining() < 10) {
                return;
            }
            int type = Short.toUnsignedInt(buffer.getShort());
            buffer.getShort();
            buffer.getInt();
            int rdlength = Short.toUnsignedInt(buffer.getShort());
            if (buffer.remaining() < rdlength) {
                return;
            }
            int dataAt = buffer.position();
            if (type == 12) {
                String target = readName(buffer, packet);
                ptr.put(name.toLowerCase(Locale.ROOT), target);
            } else if (type == 33 && rdlength >= 6) {
                buffer.getShort();
                buffer.getShort();
                int port = Short.toUnsignedInt(buffer.getShort());
                String target = readName(buffer, packet);
                srv.put(name.toLowerCase(Locale.ROOT), new Srv(target, port));
            } else if (type == 1 && rdlength == 4) {
                int b1 = Byte.toUnsignedInt(buffer.get());
                int b2 = Byte.toUnsignedInt(buffer.get());
                int b3 = Byte.toUnsignedInt(buffer.get());
                int b4 = Byte.toUnsignedInt(buffer.get());
                addr.put(name.toLowerCase(Locale.ROOT), b1 + "." + b2 + "." + b3 + "." + b4);
            }
            buffer.position(dataAt + rdlength);
            if (nameAt == buffer.position()) {
                return;
            }
        }
    }

    private static String readName(ByteBuffer buffer, byte[] packet) {
        StringBuilder name = new StringBuilder();
        int hops = 0;
        int end = -1;
        while (hops++ < 16 && buffer.hasRemaining()) {
            int len = Byte.toUnsignedInt(buffer.get());
            if (len == 0) {
                break;
            }
            if ((len & 0xC0) == 0xC0) {
                if (!buffer.hasRemaining()) {
                    break;
                }
                int offset = ((len & 0x3F) << 8) | Byte.toUnsignedInt(buffer.get());
                if (end < 0) {
                    end = buffer.position();
                }
                if (offset < 0 || offset >= packet.length) {
                    break;
                }
                buffer.position(offset);
                continue;
            }
            if (buffer.remaining() < len) {
                break;
            }
            if (name.length() > 0) {
                name.append('.');
            }
            byte[] label = new byte[len];
            buffer.get(label);
            name.append(new String(label, StandardCharsets.UTF_8));
        }
        if (end >= 0) {
            buffer.position(end);
        }
        return name.toString();
    }

    private static void skipName(ByteBuffer buffer) {
        while (buffer.hasRemaining()) {
            int len = Byte.toUnsignedInt(buffer.get());
            if (len == 0) {
                return;
            }
            if ((len & 0xC0) == 0xC0) {
                if (buffer.hasRemaining()) {
                    buffer.get();
                }
                return;
            }
            if (buffer.remaining() < len) {
                buffer.position(buffer.limit());
                return;
            }
            buffer.position(buffer.position() + len);
        }
    }

    private static byte[] encodePtrQuery(String name) {
        ByteBuffer buffer = ByteBuffer.allocate(512);
        buffer.putShort((short) 0);
        buffer.putShort((short) 0);
        buffer.putShort((short) 1);
        buffer.putShort((short) 0);
        buffer.putShort((short) 0);
        buffer.putShort((short) 0);
        for (String label : name.split("\\.")) {
            byte[] bytes = label.getBytes(StandardCharsets.UTF_8);
            buffer.put((byte) bytes.length);
            buffer.put(bytes);
        }
        buffer.put((byte) 0);
        buffer.putShort((short) 12);
        buffer.putShort((short) 1);
        byte[] query = new byte[buffer.position()];
        buffer.flip();
        buffer.get(query);
        return query;
    }

    private static String applicationFor(String ptr) {
        String value = ptr == null ? "" : ptr.toLowerCase(Locale.ROOT);
        if (value.contains("_https._tcp")) {
            return PeerServiceDiscovery.APPLICATION_HTTPS;
        }
        if (value.contains("_http._tcp")) {
            return PeerServiceDiscovery.APPLICATION_HTTP;
        }
        if (value.contains("_ssh._tcp")) {
            return PeerServiceDiscovery.APPLICATION_SSH;
        }
        return PeerServiceDiscovery.APPLICATION_UDP;
    }

    private static String instanceName(String srvName, String application) {
        String value = srvName == null ? application : srvName;
        int dot = value.indexOf('.');
        if (dot > 0) {
            value = value.substring(0, dot);
        }
        if (value.length() > PeerServiceDiscovery.MAX_NAME_LENGTH) {
            value = value.substring(0, PeerServiceDiscovery.MAX_NAME_LENGTH);
        }
        return value.isBlank() ? application : value;
    }

    private record Srv(String target, int port) {
    }
}

File: implementations/csharp/client/src/Specus.Client/PeerMesh/PeerMdnsBrowser.cs
using System.Net;
using System.Net.Sockets;
using System.Text;

namespace Specus.Client.PeerMesh;

internal static class PeerMdnsBrowser
{
    private static readonly string[] Queries =
        ["_http._tcp.local", "_https._tcp.local", "_ssh._tcp.local", "_udp.local"];

    public static IReadOnlyList<PeerMdnsCandidate> Browse(TimeSpan timeout)
    {
        try
        {
            using var socket = new UdpClient();
            socket.Client.ReceiveTimeout = Math.Max(50, (int)timeout.TotalMilliseconds);
            var mdns = new IPEndPoint(IPAddress.Parse("224.0.0.251"), 5353);
            foreach (var name in Queries)
            {
                var query = EncodePtrQuery(name);
                socket.Send(query, query.Length, mdns);
            }
            var packets = new List<byte[]>();
            var deadline = DateTime.UtcNow.Add(timeout);
            while (DateTime.UtcNow < deadline)
            {
                try
                {
                    IPEndPoint? remote = null;
                    var data = socket.Receive(ref remote!);
                    packets.Add(data);
                }
                catch
                {
                    break;
                }
            }
            return Parse(packets);
        }
        catch
        {
            return [];
        }
    }

    internal static IReadOnlyList<PeerMdnsCandidate> Parse(IReadOnlyList<byte[]> packets)
    {
        var ptr = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        var srv = new Dictionary<string, (string Target, int Port)>(StringComparer.OrdinalIgnoreCase);
        var addr = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (var packet in packets)
        {
            ParsePacket(packet, ptr, srv, addr);
        }
        var found = new Dictionary<string, PeerMdnsCandidate>(StringComparer.Ordinal);
        foreach (var (query, instance) in ptr)
        {
            if (!srv.TryGetValue(instance, out var record))
            {
                continue;
            }
            var host = addr.GetValueOrDefault(record.Target, record.Target);
            if (!IsLocalHost(host))
            {
                continue;
            }
            var application = ApplicationFor(query);
            var transport = application == "udp" ? "udp" : "tcp";
            var name = instance.Split('.', 2)[0];
            var key = $"{application}:{host}:{record.Port}";
            found[key] = new PeerMdnsCandidate
            {
                Name = name,
                Transport = transport,
                Application = application,
                TargetHost = host,
                TargetPort = record.Port,
            };
        }
        return found.Values.ToArray();
    }

    private static bool IsLocalHost(string host)
    {
        if (string.Equals(host, "localhost", StringComparison.OrdinalIgnoreCase)
            || host is "127.0.0.1" or "::1")
        {
            return true;
        }
        if (!IPAddress.TryParse(host, out var ip) || ip is null)
        {
            return false;
        }
        if (IPAddress.IsLoopback(ip))
        {
            return true;
        }
        var bytes = ip.GetAddressBytes();
        return bytes.Length == 4 && (bytes[0] == 10
            || (bytes[0] == 172 && bytes[1] is >= 16 and <= 31)
            || (bytes[0] == 192 && bytes[1] == 168)
            || (bytes[0] == 169 && bytes[1] == 254));
    }

    private static string ApplicationFor(string ptr)
    {
        var value = ptr.ToLowerInvariant();
        if (value.Contains("_https._tcp"))
        {
            return "https";
        }
        if (value.Contains("_http._tcp"))
        {
            return "http";
        }
        if (value.Contains("_ssh._tcp"))
        {
            return "ssh";
        }
        return "udp";
    }

    private static byte[] EncodePtrQuery(string name)
    {
        var buffer = new List<byte> { 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0 };
        foreach (var label in name.Split('.'))
        {
            var bytes = Encoding.UTF8.GetBytes(label);
            buffer.Add((byte)bytes.Length);
            buffer.AddRange(bytes);
        }
        buffer.Add(0);
        buffer.Add(0);
        buffer.Add(12);
        buffer.Add(0);
        buffer.Add(1);
        return [.. buffer];
    }

    private static void ParsePacket(byte[] packet, Dictionary<string, string> ptr,
        Dictionary<string, (string Target, int Port)> srv, Dictionary<string, string> addr)
    {
        if (packet.Length < 12)
        {
            return;
        }
        var offset = 12;
        var questions = (packet[4] << 8) | packet[5];
        var answers = (packet[6] << 8) | packet[7];
        var authority = (packet[8] << 8) | packet[9];
        var additional = (packet[10] << 8) | packet[11];
        for (var i = 0; i < questions && offset < packet.Length; i++)
        {
            SkipName(packet, ref offset);
            offset += 4;
        }
        var records = answers + authority + additional;
        for (var i = 0; i < records && offset + 10 <= packet.Length; i++)
        {
            var name = ReadName(packet, ref offset);
            if (offset + 10 > packet.Length)
            {
                return;
            }
            var type = (packet[offset] << 8) | packet[offset + 1];
            var rdlength = (packet[offset + 8] << 8) | packet[offset + 9];
            offset += 10;
            if (offset + rdlength > packet.Length)
            {
                return;
            }
            if (type == 12)
            {
                var dataAt = offset;
                ptr[name] = ReadName(packet, ref offset);
                offset = dataAt;
            }
            else if (type == 33 && rdlength >= 6)
            {
                var port = (packet[offset + 4] << 8) | packet[offset + 5];
                var dataAt = offset;
                offset += 6;
                srv[name] = (ReadName(packet, ref offset), port);
                offset = dataAt;
            }
            else if (type == 1 && rdlength == 4)
            {
                addr[name] = $"{packet[offset]}.{packet[offset + 1]}.{packet[offset + 2]}.{packet[offset + 3]}";
            }
            offset += rdlength;
        }
    }

    private static string ReadName(byte[] packet, ref int offset)
    {
        var labels = new List<string>();
        var hops = 0;
        var end = -1;
        while (hops++ < 16 && offset < packet.Length)
        {
            var len = packet[offset];
            if (len == 0)
            {
                offset++;
                break;
            }
            if ((len & 0xC0) == 0xC0)
            {
                if (offset + 1 >= packet.Length)
                {
                    break;
                }
                if (end < 0)
                {
                    end = offset + 2;
                }
                offset = ((len & 0x3F) << 8) | packet[offset + 1];
                continue;
            }
            offset++;
            if (offset + len > packet.Length)
            {
                break;
            }
            labels.Add(Encoding.UTF8.GetString(packet, offset, len));
            offset += len;
        }
        if (end >= 0)
        {
            offset = end;
        }
        return string.Join('.', labels);
    }

    private static void SkipName(byte[] packet, ref int offset)
    {
        while (offset < packet.Length)
        {
            var len = packet[offset];
            if (len == 0)
            {
                offset++;
                return;
            }
            if ((len & 0xC0) == 0xC0)
            {
                offset += 2;
                return;
            }
            offset += 1 + len;
        }
    }
}

internal sealed class PeerMdnsCandidate
{
    public string Name { get; set; } = "";
    public string Transport { get; set; } = "tcp";
    public string Application { get; set; } = "tcp";
    public string TargetHost { get; set; } = "";
    public int TargetPort { get; set; }
}
